lib/perl/Genome/InstrumentData/AlignmentResult.pm

```diff
@@ -1628,10 +1643,144 @@ sub verify_alignment_data {
+# This must be provided an allocation into which the bam will go (so that it can be cleaned up in the parent process once it is done).
+# The calling process is responsible for cleaning up the allocation after we are done with it.
+sub recreated_alignment_bam_file_paths {
+    my $self = shift;
+    my %p = Params::Validate::validate(@_, {disk_allocation => { isa => 'Genome::Disk::Allocation'}});
+
+    my @bams = $self->alignment_bam_file_paths;
+    if (@bams) {
+        return @bams;
+    }
+
+    my $recreated_bam = File::Spec->join($p{disk_allocation}->absolute_path, 'all_sequences.bam');
+    my $merged_bam    = $self->get_merged_bam_to_revivify_per_lane_bam;
+
+    unless ($merged_bam and -s $merged_bam) {
+        die $self->error_message('Failed to get valid merged bam to recreate per lane bam.');
+    }
+
+    my $cmd = Genome::InstrumentData::AlignmentResult::Command::RecreatePerLaneBam->create(
+        merged_bam          => $merged_bam,
+        per_lane_bam        => $recreated_bam,
+        instrument_data_id  => $self->instrument_data_id,
+        samtools_version    => $self->samtools_version,
+        picard_version      => $self->picard_version,
+        bam_header          => $self->bam_header_path,
+        comparison_flagstat => $self->flagstat_path,
+    );
+
+    unless ($cmd->execute) {
+        die $self->error_message('Failed to execute RecreatePerLaneBam');
+    }
+
+    if (-s $recreated_bam) {
+        return ($recreated_bam);
+    }
+    else {
+        die $self->error_message("After running RecreatePerLaneBam, no per-lane bam (%s) exists still!", $recreated_bam);
+    }
+}
+
+
+sub bam_header_path {
+    return File::Spec->join(shift->output_dir, 'all_sequences.bam.header');
+}
+
+
+sub flagstat_path {
+    return File::Spec->join(shift->output_dir, 'all_sequences.bam.flagstat');
+}
+
+
+sub get_merged_bam_to_revivify_per_lane_bam {
+    my $self = shift;
+    my $merged_result = $self->get_smallest_merged_alignment_result($self->get_unarchived_merged_alignment_results);
+
+    unless ($merged_result) {
+        $merged_result = $self->get_smallest_merged_alignment_result($self->get_merged_alignment_results);
+        $merged_result->_auto_unarchive;
+    }
+
+    my $merged_bam = $merged_result->merged_alignment_bam_path;
+    unless (-s $merged_bam) {
+        die $self->error_message("Merged bam (%s) does not exist for merged result id (%s)", $merged_bam, $merged_result->id);
+    }
+    return $merged_bam;
+}
+
+sub get_merged_alignment_results {
+    my $self = shift;
+    # Always load from the database, since other merged results may have committed since we updated the UR cache
+    my @results = Genome::InstrumentData::AlignmentResult::Merged->load(
+        'inputs.value_id' => $self->instrument_data_id,
+        test_name => $self->test_name,
+    );
+    my @filtered_results = $self->filter_non_database_objects(@results);
+    return $self->filter_non_matching_results(@filtered_results);
```

Comment:
The reasons why we do a reverse_lookup:
- Merged.pm already knows how to get its alignment results, but the logic is difficult to reverse and generalize and considers both test_name and parameters, as well as instrument data id
- Getting every merge that shares an instrument data id with me, and then letting Merge.pm do the reverse lookup solves this

So I think the 'pro' is that we don't have to invert and duplicate code. The 'con' is that it's confusing and inefficient (but probably runs fast enough that we do not care). I'm not convinced this is the very best solution, it's just what we could come up with so far. I think the best solution I can think of is to have a stable SoftwareResult::User relationship between per-lane and merged alignment results.